cachetory/decorators/shared.py

```diff
@@ -23,3 +26,29 @@ def make_default_hashed_key(callable_: Callable[..., Any], *args: Any, **kwargs:
     Uses `blake2s` as the fastest algorithm from `hashlib`.
     """
     return blake2s(make_default_key(callable_, *args, **kwargs).encode()).hexdigest()
+
+
+T = TypeVar("T")
+
+
+async def call_sync_or_async(
+    fun: Union[
+        Callable[..., T],
+        Callable[..., Awaitable[T]],
```

Comment:
This does not work at all, because of course `T` could simply be an `Awaitable`. mypy chokes pretty hard when working with `maybe sync maybe async`, and I do too.

At this point I am kinda lost, because I feel I am writing more technical debt than actual code. So I leave it up to you @eigenein 😆 Maybe just inlining the `if inspect.iscoroutinefunction(func)` directly into `async_.py` would be best? Like so:

```python
def cached(
    cache: Union[
        Cache[ValueT, WireT],
        Callable[..., Cache[ValueT, WireT]],
        Callable[..., Awaitable[Cache[ValueT, WireT]]],
    ],  # no way to use `P` here
    *,
    make_key: Callable[..., str] = shared.make_default_key,  # no way to use `P` here
    time_to_live: Optional[Union[timedelta, Callable[..., timedelta], Callable[..., Awaitable[timedelta]]]] = None,
    if_not_exists: bool = False,
) -> Callable[[Callable[P, Awaitable[ValueT]]], Callable[P, Awaitable[ValueT]]]:

    def wrap(callable_: Callable[P, Awaitable[ValueT]]) -> Callable[P, Awaitable[ValueT]]:
        @wraps(callable_)
        async def cached_callable(*args: P.args, **kwargs: P.kwargs) -> ValueT:
            if callable(cache):
               cache_ = await cache(_callable, *args, **kwargs) if inspect.iscoroutinefunction(cache) else  cache(_callable, *args, **kwargs)
            else:
                cache_ = cache
            key_ = make_key(callable_, *args, **kwargs)
            time_to_live_ = (
                await shared.call_sync_or_async(time_to_live, key=key_) if callable(time_to_live) else time_to_live
            )
            if callable(time_to_live):
                time_to_live_ = await time_to_live(key=key_) if inspect.iscoroutinefunction(time_to_live) else  time_to_live(key=key_)
            else:
                time_to_live_ = time_to_live

            value = await cache_.get(key_)
            if value is None:
                value = await callable_(*args, **kwargs)
                await cache_.set(key_, value, time_to_live=time_to_live_, if_not_exists=if_not_exists)
            return value

        return cached_callable

    return wrap
```

I simply don't know how to write super generic stuff in python with types working 😅

Reply:
> This does not work at all, because of course T could simply be an Awaitable. mypy chokes pretty hard when working with maybe sync maybe async, and I do too.

I must admit, the typing system in Python/Mypy isn't as mature as expected. Let's just inline it for now since we only have 2 flavours at the moment, and worry about it later (and then maybe Mypy is fixed)

Reply:
done! @eigenein .

mypy still chokes (`reveal_type` is `Any`). Vscode also chokes (`type` is `Any | `Awaitable[timedelta]` | `timedelta`).

🤷🏽‍♂️ you cannot win it seems. `mypy` does not complain anyway, but that's a low bar.

Reply:
@JeroennC kindly offered a solution for this, which as you can see involved `@overloads` and (very importantly), changing the order of the `ifelse` in `async_`. Now, `reveal_type` for both `cache_` and `time_to_live_` are correct ✅